include/ozo/detail/deadline.h

```diff
@@ -33,32 +38,44 @@ class deadline_handler {
 
     using allocator_type = asio::associated_allocator_t<Continuation>;
 
-    allocator_type get_allocator() const noexcept { return asio::get_associated_allocator(handler_);}
+    allocator_type get_allocator() const noexcept { return asio::get_associated_allocator(ctx_->handler_);}
 
 private:
-    template <typename Target>
+
+    timer_type timer_;
+    executor_type executor_;
+    struct context {
+        Continuation handler_;
+        TimerHandler on_deadline_;
+        std::atomic<int> state_{2};
+
+        context(Continuation&& handler, TimerHandler&& on_deadline)
+        : handler_(std::move(handler)), on_deadline_(std::move(on_deadline)) {
+        }
+    };
+
+    std::shared_ptr<context> ctx_;
+
     struct wrapper {
         using executor_type = deadline_handler::executor_type;
 
         executor_type get_executor() const noexcept { return executor_;}
 
-        using allocator_type = asio::associated_allocator_t<Target>;
+        using allocator_type = deadline_handler::allocator_type;
 
-        allocator_type get_allocator() const noexcept { return asio::get_associated_allocator(target_);}
+        allocator_type get_allocator() const noexcept { return allocator_;}
 
         void operator() (error_code ec) {
-            if (ec != asio::error::operation_aborted) {
-                asio::dispatch(ozo::detail::bind(std::move(target_), std::move(ec)));
+            if (--(ctx_->state_) && ec != asio::error::operation_aborted) {
```

Comment:
If ec = asio::error::operation_aborted handler will never be called because state was decremented. It expected?

Reply:
Yep, thanks.
**Done**